Report timing

MartinDelille opened the pull request:
This PR aims to fix #148 but is still a work in progress.

I just moved the timing computation to `progress_bar.h` and wanted to know if it looked fine for you?

MartinDelille commented:
@sbeyer I add the `--report-timing` option (basing myself on the `--dry-run` option) but it currently doing nothing.

I'm not really sure how to pass this option to the reporter. Through the `run_policies`? through the `controller`? directly to the `reporter`? Through one of the `interface`?

A design issue is that this option should be default to:
- `true` for the `xunit` reporter
- `false` otherwise

I'm not sure how to handle that.

sbeyer commented:
> I just moved the timing computation to `progress_bar.h` and wanted to know if it looked fine for you?

If you would have asked me in advance, I would have said that the timing computation goes to a separate class (separation of concerns), but seeing `progress_base.h` in its entirety, I think it's okay to have it there. Thank you.

Nitpicking: I prefer the standard commit message format where the first letter is upper-cased (as you did in your second commit) :smile: 

> I'm not really sure how to pass this option to the reporter. Through the `run_policies`? through the `controller`? directly to the `reporter`? Through one of the `interface`?

It's interesting how complicated the architecture of bandit is already. :smile:
The run policies define "what tests should be run" (and also: "how" since `--break-on-failure` and `--dry-run` is handled, but this will change before version 3.0); the controller is the "main" thing where all is plugged together (and it is configurable, e.g., you can say *which* run policy, formatter, reporter, etc. are plugged together); the reporter manages the way things are presented to the user. "Timing" doesn't really fit in any class of this, but we currently think of it as time measurement is something the reporter does, so you should pass it directly to the constructors of the reporters that can handle this option (in `runner.h` method `use_default_reporters()`).

> A design issue is that this option should be default to:
> 
>     * `true` for the `xunit` reporter
> 
>     * `false` otherwise

That would be a bad design decision. It should always default to `false`. This will, by default, make the `xunit` reporter return `0.0` as it has been before commit 8704dc233510abab71f9fd58a846a1b94085ab23.

sbeyer commented on bandit/options.h:
```diff
@@ -93,6 +93,9 @@ namespace bandit {
                     "\tSkip all tests. Use to list available tests"},
+                {REPORT_TIMING, 0, "", "report-timing", argument::None,
+                    "  --report-timing, "
+                    "\tAdd timing information to info reporter"},
```
This should probably be something like **"Instruct reporter to report timing information"** (assuming that not every reporter will handle it).

MartinDelille replied:
Ok changed.

MartinDelille commented:
Thank you for this detailed answer!

> If you would have asked me in advance, I would have said that the timing computation goes to a separate class (separation of concerns), but seeing progress_base.h in its entirety, I think it's okay to have it there. Thank you.

You mean by creating a new class like `time_base` (or any naming you'll find more suitable) that would inherit `progress_base` and be inherited by `colored_base` and `xunit`? If yes I can definitely change that!

> Nitpicking: I prefer the standard commit message format where the first letter is upper-cased (as you did in your second commit) 😄 

No problem!

> "Timing" doesn't really fit in any class of this, but we currently think of it as time measurement is something the reporter does, so you should pass it directly to the constructors of the reporters that can handle this option (in runner.h method use_default_reporters()).

Ok  that's what I started was thinking about and I think I found a way to transmit the `report_timing` parameter.

>> A design issue is that this option should be default to:
>> * `true` for the `xunit` reporter
>> * `false` otherwise

> That would be a bad design decision. It should always default to false. This will, by default, make the xunit reporter return 0.0 as it has been before commit 8704dc2.

I was just worrying about all current bandit xunit reporter user will need to add the `--report-timing` option in the future but if you're fine with this I totally agree that this would be the best design decision.

MartinDelille commented:
What do you think of this first implementation?

It currently only show timing per test. No total duration is displayed (which would make sense for `dots` reporter for example) but that would no be hard to add (and would do it in this PR).

I doesn't handle timing per context which would imply to store context start timing in the stacks.

I need to add test for the `xunit` reporter.

MartinDelille commented:
I'm wondering what is going wrong with this gigantic duration in the failing and errored case because I don't reproduce (I'm using clang).

sbeyer commented on bandit/reporters/info.h:
```diff
@@ -274,6 +281,17 @@ namespace bandit {
+        if (report_timing_) {
+          std::chrono::duration<double> dur_in_sec(testcase_duration_);
+          stm_
+            << colorizer_.neutral()
```
Not sure if this is the right color choice (well, that's about taste, I guess), so I will have to check later when I check out your PR branch to test it myself. :-)

MartinDelille replied:
Here are some screenshots:
<img width="581" alt="Screenshot 2019-10-27 15 40 52" src="https://user-images.githubusercontent.com/879889/67636395-6e5d3900-f8d0-11e9-9895-1a8ea9e275a7.png">
<img width="720" alt="Screenshot 2019-10-27 15 42 09" src="https://user-images.githubusercontent.com/879889/67636397-73ba8380-f8d0-11e9-86ba-9a0229f73fe7.png">

sbeyer replied:
Your color settings are very non-standard in some way "[ PASS ]"and "Passed: <number>" is usually green.

MartinDelille replied:
Yes indeed. That is why it misled me in this issue: https://github.com/banditcpp/bandit/issues/149

sbeyer commented on bandit/reporters/info.h:
```diff
@@ -274,6 +281,17 @@ namespace bandit {
+          stm_
+            << colorizer_.neutral()
+            << " ("
+            << std::to_string(dur_in_sec.count())
```
I'll have to check this later, too, because I am not at all fluent in `std::chrono`. I know that it's just the code taken from the xunit reporter but what I *would* expect is the number of seconds but with precision in milliseconds, i.e. "0.123 seconds", not "0 seconds", not "0.12333333 seconds".

MartinDelille replied:
Fixed here: 4e9ef0f

Has you can see I set the format through [std::fixed](http://www.cplusplus.com/reference/ios/fixed/). Is it fine for you?

sbeyer commented on bandit/reporters/progress_base.h:
```diff
@@ -94,6 +100,11 @@ namespace bandit {
 
     protected:
+      void update_test_duration() {
+        testcase_duration_  = std::chrono::high_resolution_clock::now() - testcase_start_time_point_;
```
This is just again nitpicking about style: can you change the double space before `=` to a single space? (No alignment)

MartinDelille replied:
Sorry I missed that and will fix it in my next commit. I'm personnaly using [uncrustify](http://uncrustify.sourceforge.net/) to format file automatically and a job to the CI to check files are well formated.

I saw a [clang-format configuration](https://github.com/banditcpp/snowhouse/blob/master/.clang-format) in the latest snowhouse master branch. Adding that to bandit would help!

sbeyer commented on bandit/reporters/xunit.h:
```diff
@@ -101,16 +94,15 @@ namespace bandit {
+          std::chrono::duration<double> dur_in_sec(testcase_duration_);
+          work_stm_ << " time=\"" << std::to_string(dur_in_sec.count());
+        }
+        work_stm_ << "\">\n";
```
According to the XML specification(*), `time` is a required attribute, so I'd prefer if it says `time="0.0"` if the feature is disabled. This is also the behavior of former bandit versions.

(*) I am ignoring the fact that we don't comply to the specification regarding other things, but... well :sweat_smile:

MartinDelille replied:
Fixed here: 47232d5

What about displaying a warning a warning on `std::clog` or `std::cerr` if reporter `xunit` is specified without `--report-timing`?

MartinDelille replied:
Other question: if we display always the `time` attribute for `xunit` reporter, should I use [std::thread::sleep_for](http://www.cplusplus.com/reference/thread/this_thread/sleep_for/) to avoid zero value?

sbeyer replied:
I have also thought about that but I think no, we just assume zero time in the tests, although this could mean that tests fail under big load.

sbeyer commented on bandit/runner.h:
```diff
@@ -65,22 +65,22 @@ namespace bandit {
     });
     choices.reporters.add("dots", [&](detail::controller_t& controller) {
-      controller.set_reporter(new bandit::reporter::dots(controller.get_formatter(), controller.get_colorizer()));
+      controller.set_reporter(new bandit::reporter::dots(controller.get_formatter(), controller.get_report_timing(), controller.get_colorizer()));
```
According to comment :abc:, it's only necessary to change the reporter invocations that support the report of timing.

MartinDelille replied:
Fixed here: 043991d

sbeyer commented on docs/runningtests.md:
```diff
@@ -150,6 +151,13 @@ This is the default.
 
+### Timing
+
+#### `--report-timing
```
You forgot a backquote at the end.

MartinDelille replied:
Fixed here: a32d962

sbeyer commented on specs/reporters/info.cpp:
```diff
@@ -376,6 +376,43 @@ go_bandit([]() {
+        reporter->test_run_starting();
+        reporter->test_run_starting();
+        reporter->context_starting("context");
+        reporter->it_starting("report timing");
```
I'm grammar pedantic: "reports timing"

MartinDelille replied:
Fixed here: a32d962

sbeyer commented on specs/reporters/crash.cpp:
```diff
@@ -10,7 +10,7 @@ go_bandit([]() {
     before_each([&]() {
       stm.str(std::string());
-      reporter.reset(new reporter::crash(stm, formatter));
+      reporter.reset(new reporter::crash(stm, formatter, false));
```
As mentioned in :abc:, this should not be necessary because the reporter does not support timing reports.

MartinDelille replied:
Fixed here: 043991d

sbeyer commented on specs/reporters/info.cpp:
```diff
@@ -376,6 +376,43 @@ go_bandit([]() {
+        reporter->context_ended("context");
+        reporter->test_run_complete();
+
+        AssertThat(output(), Contains("[ FAIL ] it fails (0.0"));
```
Since I am only checking coarsely, I don't *know* why the CI fails here (and below), but it looks like a problem with uninitialized values.

sbeyer commented:
> I'm wondering what is going wrong with this gigantic duration in the failing and errored case because I don't reproduce (I'm using clang).

Can you enable the undefined-behavior sanitizer (UBSan) of clang and check again?

MartinDelille commented:
> Can you enable the undefined-behavior sanitizer (UBSan) of clang and check again?

I'm not very familiar with `cmake` but I added this to the `CMakeLists.txt`:

```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=undefined")
```

The tests pass without error on my mac.

MartinDelille commented on bandit/reporters/progress_base.h:
```diff
@@ -11,9 +11,9 @@
+      progress_base(const detail::failure_formatter_t& formatter, bool report_timing)
           : specs_run_(0), specs_succeeded_(0), specs_failed_(0), specs_skipped_(0),
-            failure_formatter_(formatter) {}
+            failure_formatter_(formatter), report_timing_(report_timing) {}
```
Fixed here: 043991d

MartinDelille commented on bandit/reporters/progress_base.h:
```diff
@@ -116,6 +116,7 @@ namespace bandit {
       std::chrono::high_resolution_clock::time_point testcase_start_time_point_;
       std::chrono::duration<double> testcase_duration_;
       std::chrono::nanoseconds testsuite_runtime_;
+      bool report_timing_;
```
Fixed here: 043991d

MartinDelille commented on bandit/reporters/crash.h:
```diff
@@ -8,11 +8,11 @@
-      crash(const detail::failure_formatter_t& formatter)
-          : crash(std::cout, formatter) {}
+      crash(const detail::failure_formatter_t& formatter, bool report_timing)
+          : crash(std::cout, formatter, report_timing) {}
```
Fixed here: 043991d

MartinDelille commented:
Any idea how to solve this CI issue?

sbeyer commented:
> Any idea how to solve this CI issue?

Not yet. I just pulled your PR to check in my own setup (g++ 9 with UBSan and ASan activated), and I get "-0.000 seconds" where the AppVeyor gets "-6277438562204192487878988888393020692503707483087375482269988814848.000 seconds". Not quite sure what's happening at the moment. Will look later.

MartinDelille commented:
Ok I will try to set up a windows machine to try to reproduce the bug.

sbeyer commented:
> Ok I will try to set up a windows machine to try to reproduce the bug.

As I said, it is also weird with my setup under Linux, so I guess as long as this problem is not solved, there is no need to setup a Windows machine. Here, the debugger says
```
Breakpoint 1, bandit::reporter::progress_base::update_test_duration (this=0x60d0000002b0)
    at /home/sbeyer/src/banditcpp/bandit/bandit/reporters/progress_base.h:104
104	        testcase_duration_ = std::chrono::high_resolution_clock::now() - testcase_start_time_point_;
(gdb) p testcase_duration_ 
$6 = {__r = -1.8325506472120096e-06}
```
i.e., that the time difference is negative, that's why I get the `-0.000 seconds` output. I wonder why this happens, and moreover, why this happens _reproducible every single time_. Some type problem? Using `std::chrono::milliseconds` (integer type instead of double) did not change anything; and the [example on cppreference](http://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/chrono/high_resolution_clock/now.html) is just the same, only with `auto`. I'm puzzled.

sbeyer commented:
A heads-up: Although my sanitizers did not notice any uninitialized values, the static analysis by Coverity Scan did (on the master branch in the xunit code). I will check if this is a true positive and if a fix for that might solve our problem here.

sbeyer commented:
Okay, I totally **don't** understand this, but explicitly calling the default constructor for `testcase_duration_` solves the problem on my machine. So, could you try this change?
```diff
diff --git a/bandit/reporters/progress_base.h b/bandit/reporters/progress_base.h
index f4b26f3..c3de618 100644
--- a/bandit/reporters/progress_base.h
+++ b/bandit/reporters/progress_base.h
@@ -13,7 +13,8 @@ namespace bandit {
     struct progress_base : public interface {
       progress_base(const detail::failure_formatter_t& formatter)
           : specs_run_(0), specs_succeeded_(0), specs_failed_(0), specs_skipped_(0),
-            failure_formatter_(formatter) {}
+            failure_formatter_(formatter),
+            testcase_start_time_point_(), testcase_duration_(), testsuite_runtime_() {}
 
       progress_base& operator=(const progress_base&) {
         return *this;
```

MartinDelille commented:
It pass! I don't recall what the problem was exactly but I know that the advantage of using clang compiler is to avoid this kind of annoying problems.

sbeyer commented:
I checked again and spotted the bad guy. It is unrelated to your patch. The bad guy is the info reporter: overridden methods do not always call the respective `progress_base` base method, hence the timing updates are not performed. I will fix this in a separate branch. I will also give you a further review; then you could rebase on top of current master and we can finally merge this fine work. Thank you very much!

MartinDelille commented:
Thank you again for the awesome work your doing on this project!

sbeyer commented on specs/reporters/xunit.cpp:
```diff
@@ -12,7 +12,8 @@ go_bandit([]() {
       stm.str(std::string());
-      reporter.reset(new reporter::xunit(stm, formatter));
+      reporter.reset(new reporter::xunit(stm, formatter, true));
+      // TODO: do test without timing
```
Please do, or don't (should contain `time="0` in both cases), but please don't add a TODO ;)

MartinDelille commented:
Ok I fixed the problem and rebased as you requested. I sometime keep intermediate commit before merge it helps during development to keep a track on fix, but I'm totally fine with keeping a clean history before merging.

Don't hesitate if you have more feedback!

sbeyer commented on bandit/reporters/progress_base.h:
```diff
@@ -92,6 +100,18 @@ namespace bandit {
+        testsuite_runtime_ += std::chrono::duration_cast<std::chrono::nanoseconds>(testcase_duration_);
+      }
+
+      std::string print_time(double t) {
```
The name of the method does not represent what it is doing (i.e., it does not print anything).

MartinDelille replied:
time_to_string?

sbeyer commented on bandit/reporters/progress_base.h:
```diff
@@ -92,6 +100,18 @@ namespace bandit {
+      std::string print_time(double t) {
+        std::stringstream ss;
+        ss.precision(3);
+        ss << std::fixed << t;
```
Okay it seems you did not like the rounding approach I had in mind. Your solution always shows 3 digits. If that is what you want (I don't care), I'm totally fine with it.

MartinDelille replied:
This is not that I didn't like it, this is that the test were failing because a duration of zero was printing `0` and `0.` was expected.

I could also have updated the test but having a fixed length output for the duration might be confortable what do you think?

sbeyer replied:
As I said, I don't care. Just keep it this way. :)

sbeyer commented on bandit/reporters/progress_base.h:
```diff
@@ -4,6 +4,8 @@
 #include <sstream>
 #include <vector>
+#include <chrono>
+#include <cmath>
```
I think `cmath` is unnecessary without `std::round`.

sbeyer commented on bandit/reporters/xunit.h:
```diff
@@ -101,16 +97,14 @@ namespace bandit {
-        std::chrono::duration<double> dur_in_sec(dur);
-        work_stm_ << " time=\"" << std::to_string(dur_in_sec.count()) << "\">\n";
+        work_stm_ << " time=\""
+                  << (report_timing_ ? testcase_duration_.count() : 0.0)
```
no `print_time()` call here?

MartinDelille replied:
Oups my bad!

sbeyer commented:
Thank you very much!

MartinDelille commented:
It was very instructive to work on this pull request ! Thank you for taking the time to help me merge it !
